lib/Dialect/FIRRTL/Transforms/Dedup.cpp

```diff
@@ -1251,6 +1251,13 @@ struct Deduper {
         targetMap[nla.getAttr()].insert(to);
         continue;
       }
+      // If the annotation is a local tracker, don't make it non-local. This
+      // allows trackers that refer to all instances when the user wants that.
+      if (anno.isClass("circt.tracker")) {
+        newAnnotations.push_back(anno);
+        continue;
+      }
+
```

Comment:
I am not totally sure how all the path stuff currently works, but the change doesn't seem totally right - `makeAnnotationNonLocal` is maybe not named that great, but it adds more context to a HierPath if it is needed, and it must be called on non-local annotations to make sure their path stays disambiguated.  How can we tell if a path is local or not? Should you make the condition `anno.isClass("circt.tracker") && !anno.hasMember("circt.nonlocal")` kind of thing? 

Does Chisel emit absolute paths to the top of the circuit? If so I could see this code working in practice, but I think someone could write an example where this is changing the behaviour the behavior. Trying to get the idea out there:
```
circuit Foo:
   public module Foo:
     inst bar0 of Bar0
     inst bar1 of Bar1
  module Bar0:
     output path : Path
     wire w
     connect path, path("Bar0>w")
   module Bar1:
     output path : Path
     wire w
     connect path, path("Bar1>w")
```
So my concert is that when Bar0 and Bar1 dedup, the path doesn't get extended with more context, and we would get a complaint about an ambiguous path (which instance of Bar0 is the path referring to?). Is something like that possible?

Reply:
> Should you make the condition anno.isClass("circt.tracker") && !anno.hasMember("circt.nonlocal") kind of thing?

Yep, this is effectively the logic here. This is checking for tracker annotations just after we check if the annotation is already non-local: https://github.com/llvm/circt/blob/2085d0d91e3b540ea765539c58bea305dad5a6aa/lib/Dialect/FIRRTL/Transforms/Dedup.cpp#L1247-L1253

> Does Chisel emit absolute paths to the top of the circuit?

Sometimes, but it is up to the user if they make a hierarchical or local target. In the case of hierarchical targets, nothing changes here, since they're already NLAs by dedup, and the logic I linked just above kicks in. So this is really about handling local targets, just like the example you shared. Specifically, Chisel has started emitting paths very similar to this: https://github.com/chipsalliance/chisel/commit/dd2f57ede872afc2e2b9c00dcdbcd397400bcdfd

I cleaned up your example slightly to make it compile, and to have LowerClasses actually do something interesting:

```
FIRRTL version 4.0.0

circuit Foo:
  public module Foo:
    output path0 : Path
    output path1 : Path
    inst bar0 of Bar0
    inst bar1 of Bar1
    propassign path0, bar0.path
    propassign path1, bar1.path

  module Bar0:
    output path : Path
    wire w : UInt<1>
    invalidate w
    propassign path, path("OMReferenceTarget:~Foo|Bar0>w")

  module Bar1:
    output path : Path
    wire w : UInt<1>
    invalidate w
    propassign path, path("OMReferenceTarget:~Foo|Bar1>w")
```

With this change, I think we're getting exactly what we'd want:

```
  om.class @Foo_Class(%basepath: !om.frozenbasepath) {
    %0 = om.frozenbasepath_create %basepath "Foo/bar0"
    %1 = om.object @Bar0_Class(%0) : (!om.frozenbasepath) -> !om.class.type<@Bar0_Class>
    ...
    %3 = om.frozenbasepath_create %basepath "Foo/bar1"
    %4 = om.object @Bar0_Class(%3) : (!om.frozenbasepath) -> !om.class.type<@Bar0_Class>
    ...
  }
  om.class @Bar0_Class(%basepath: !om.frozenbasepath) {
    %0 = om.frozenpath_create reference %basepath "Bar0>w"
    om.class.field @path, %0 : !om.frozenpath
  }

```

The local path is just referring to the deduped `Bar0>w` after Dedup runs, and LowerClasses [recently](https://github.com/llvm/circt/commit/002427065e8b0751a5c0ed8a980addf99f7ad82a) started supporting passing local paths through to the output without trying to disambiguate them or erroring in the face of multiple instantiation. Note that the different instances of `Bar0_Class` do get different base paths, so the final elaborated paths here get built up hierarchically following the instance graph.

This doesn't show up in your exact example, but in similar cases, without the change in this PR, after Bar0 and Bar1 dedupe, we'd clone the tracker into two: one for each path to the `w` wire. This will fail in LowerClasses here: https://github.com/llvm/circt/blob/2085d0d91e3b540ea765539c58bea305dad5a6aa/lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp#L715-L725


This is the general thrust of this change and some of the recent changes to paths--as we start seeing more uses of D/I in Chisel, the notion that you can even make a unique, absolute path from the top of the circuit becomes impossible, so instead we're doing more with local paths, and not forcing ourselves to disambiguate in places we did previously. Instead, paths can safely be local, and compose hierarchically using the basepath infrastructure in place in LowerClasses.

Does that help alleviate your concerns @youngar ?

Reply:
Yeah! Either I just can't read code or I forgot how this works - the context adding that I was talking about happens elsewhere and has nothing to do with your change. Sorry for the noise!